### PR TITLE
feat(dashboard): include dependency lookup warnings

### LINKED ISSUE #15651: Raise issue for "Dependency Lookup Warnings"
### What would you like Renovate to be able to do?

Renovate could raise an optional dedicated issue if any dependency could not be looked up (e.g. because authorization was missing). 

The onboarding PR already includes this section and this is very helpful to identify if all dependencies could be looked up (and potentially updated). For an already onboarded project when it later adds additional dependencies these warnings are not raised and only logged in the log file and easily missed. 

I would like to see that Renovate raises a dedicated issue for these lookup warnings to ensure that one could be aware of them.   

### If you have any ideas on how this should be implemented, please tell us here.

This could be implemented similiar to [`raiseConfigWarningIssue`](https://github.com/renovatebot/renovate/blob/50d23e6576ee1c068de5fb41a4e58cbc3059a685/lib/workers/repository/error-config.ts#L8) and per default included in `suppressNotifications` to not bug people about if they don't want to receive alerts.

### Is this a feature you are interested in implementing yourself?

Maybe

### PR DESCRIPTION
**Pull request by MaronHatoum**
## Changes
added "Dependency Lookup Warnings" to dashboard

## Context
closes https://github.com/renovatebot/renovate/issues/15651

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
Example repo link?

**Comment by MaronHatoum** [comment]
> Example repo link?

https://github.com/MaronHatoum/dashboard-dependency-lookup-warnings-reproduction

**Review by rarkins — CHANGES_REQUESTED** [review]
Current screenshot:
image

I think we need a separator between the (new) dependency warnings and the branches section. We also need to decide if we want the structure to be like this:
- dependency warnings
- branch list
- dependency list

You could argue that dependency warnings are most important so should go to the top, but on the other hand we'd be splitting the two dependency sections.

I'd like to see how this looks with:
- More than one lookup warning
- An example which includes both branches and warnings (this could be achieved using npm with some outdated packages plus some random package names which don't exist)

**Comment by MaronHatoum** [comment]
@rarkins, this repository contains the example you asked for:
- more than one lookup warning
- includes both branches and warnings

https://github.com/MaronHatoum/lookup-warnings

**Comment by rarkins** [comment]
image

What does everyone think?

**Comment by PhilipAbed** [comment]
@betterPeleg what do you think?

**Comment by betterPeleg** [comment]
Screen Shot 2022-07-05 at 11 28 30

i like it.
small changes - 1. can you add another seperator (like the one over the title over "Dependency lookup warnings") below the warnings and above the "Open" section.
2. can you change the marking of the text (for example "Failed to look up dependency deppp") to only emphasis the dependency name. (and not the whole sentence)

note - regarding the first point please let me see how it looks like before creating the PR
thanks

**Comment by MaronHatoum** [comment]
@betterPeleg this is how the dashboard will look like after the changes you asked: https://github.com/MaronHatoum/lookup-warnings/issues/16
If you have any comments please let me know 
image

**Comment by betterPeleg** [comment]
looks good to me

**Comment by MaronHatoum** [comment]
Are any more changes needed here?

**Comment by rarkins** [comment]
We can maybe go with it as-is, although I wonder if it would be better to remove the repeating of "Failed to look up dependency " on every line and instead just have a long list like:

Renovate failed to look up the following dependencies: `deppp`, `ppped`.

**Comment by betterPeleg** [comment]
yes let's do that @rarkins

**Comment by MaronHatoum** [comment]
@rarkins @betterPeleg
is this the expected result? https://github.com/MaronHatoum/lookup-warnings/issues/19 

image

**Comment by betterPeleg** [comment]
yes. looks good to me

**Review by PhilipAbed — DISMISSED** [review]

File: lib/workers/repository/index.ts [hunk 1]
```diff
@@ -59,7 +59,7 @@ export async function renovateRepository(
         }
         logger.debug(`Automerged but already retried once`);
       } else {
-        await ensureDependencyDashboard(config, branches);
+        await ensureDependencyDashboard(config, branches, packageFiles);
```
Comment (PhilipAbed) [inline lib/workers/repository/index.ts:62]: don't the packageFiles exist in the config? check it out they might xD config is a big object

Comment (MaronHatoum) [reply lib/workers/repository/index.ts:62]: as I remember that I checked if packageFiles exists in config object but it does not.

**Comment by nabeelsaabna** [comment]
@MaronHatoum please fix conflicts

File: lib/workers/repository/errors-warnings.ts [hunk 2]
```diff
@@ -82,3 +82,40 @@ export function getDepWarningsPR(
...
+  packageFiles: Record<string, PackageFile[]>
+): string {
+  const { warnings, warningFiles } = getDepWarnings(packageFiles);
+  let warningText = '';
+  if (!warnings.length) {
+    return '';
+  }
+  logger.debug(
+    { warnings, warningFiles },
+    'Found package lookup warnings in onboarding'
```
Comment (rarkins) [inline lib/workers/repository/errors-warnings.ts:96]: only in onboarding?

File: lib/workers/repository/errors-warnings.ts [hunk 3]
```diff
@@ -82,3 +82,40 @@ export function getDepWarningsPR(
...
+  );
+  warningText = emojify(
+    `\n---\n\n### :warning: Dependency Lookup Warnings :warning:\n\n`
+  );
+  let firstTime = true;
+  warnings.forEach((w) => {
+    const splitStr = w.split(' ');
+    const dep = splitStr.slice(5).join(' ');
```
Comment (rarkins) [inline lib/workers/repository/errors-warnings.ts:104]: Avoid magic numbers

**Review by rarkins — CHANGES_REQUESTED** [review]
Also needs deconflicting

File: lib/workers/repository/dependency-dashboard.ts [hunk 4]
```diff
@@ -254,6 +257,13 @@ export async function ensureDependencyDashboard(
     }
     issueBody += '\n';
   }
+
+  const warn = getDepWarningsDashboard(packageFiles);
+  if (warn.length) {
+    issueBody += warn;
```
Comment (nabeelsaabna) [inline lib/workers/repository/dependency-dashboard.ts:263]: check max issue body length/size ?

File: lib/workers/repository/errors-warnings.ts [hunk 5]
```diff
@@ -82,3 +80,38 @@ export function getDepWarningsPR(
     '\n\n';
   return warningText;
 }
+
+export function getDepWarningsDashboard(
```
Comment (nabeelsaabna) [inline lib/workers/repository/errors-warnings.ts:84]: lets simplify this function

**Review by nabeelsaabna — CHANGES_REQUESTED** [review]
check review comments

File: lib/workers/repository/dependency-dashboard.ts (hunk 4 above)
Comment (MaronHatoum) [reply lib/workers/repository/dependency-dashboard.ts:263]: after preparing the dependency dashboard body function `smartTruncate()` checks the body length and changes it if needed.

Flow:
1.
https://github.com/renovatebot/renovate/blob/6246a61bf4a4050e07632fe2658f774528c1e121/lib/workers/repository/dependency-dashboard.ts#L359-L367
2.
https://github.com/renovatebot/renovate/blob/6246a61bf4a4050e07632fe2658f774528c1e121/lib/modules/platform/github/index.ts#L1575
3.
https://github.com/renovatebot/renovate/blob/6246a61bf4a4050e07632fe2658f774528c1e121/lib/modules/platform/utils/pr-body.ts#L8-L33

File: lib/workers/repository/errors-warnings.ts (hunk 5 above)
Comment (nabeelsaabna) [reply lib/workers/repository/errors-warnings.ts:83]: much better 💯

**Review by nabeelsaabna — DISMISSED** [review]

File: lib/workers/repository/errors-warnings.ts [hunk 6]
```diff
@@ -82,3 +79,29 @@ export function getDepWarningsPR(
...
+  const { warnings, warningFiles } = getDepWarnings(packageFiles);
+  if (!warnings.length) {
+    return '';
+  }
+
+  let depWarnings = '';
+  warnings.forEach((w) => {
+    const dep = w.replace('Failed to look up dependency ', '');
+    depWarnings += '`' + dep + '`, ';
+  });
```
Comment (rarkins) [inline lib/workers/repository/errors-warnings.ts:95]:
```suggestion
  const depWarnings = warnings
    .map((w) => w.replace('Failed to look up dependency ', ''))
    .map((dep) => '`' + dep + '`')
    .join(', ');
```

What do you think about this change? Then doesn't need trailing slash removed later.

Comment (MaronHatoum) [reply lib/workers/repository/errors-warnings.ts:95]: you are right, i will push the changes
thanks

Comment (nabeelsaabna) [reply lib/workers/repository/errors-warnings.ts:95]: nice 👍

File: lib/workers/repository/errors-warnings.ts [hunk 7]
```diff
@@ -82,3 +79,28 @@ export function getDepWarningsPR(
...
+    .map((w) => w.replace('Failed to look up dependency ', ''))
+    .map((dep) => '`' + dep + '`')
+    .join(', ');
+
+  logger.debug({ warnings, warningFiles }, 'Found package lookup warnings');
+  let warningText = emojify(
+    `\n---\n\n### :warning: Dependency Lookup Warnings :warning:\n\n`
+  );
+  warningText += `-   Renovate failed to look up the following dependencies: `;
+  warningText += depWarnings.substring(0, depWarnings.length); // append after removing the last comma
```
Comment (MaronHatoum) [inline lib/workers/repository/errors-warnings.ts:101]:
```suggestion
  warningText += depWarnings;
```

File: lib/workers/repository/dependency-dashboard.ts [hunk 8]
```diff
@@ -256,6 +259,13 @@ export async function ensureDependencyDashboard(
     }
     issueBody += '\n';
   }
+
+  const warn = getDepWarningsDashboard(packageFiles);
+  if (warn.length) {
```
Comment (nabeelsaabna) [inline lib/workers/repository/dependency-dashboard.ts:264]:
```suggestion
  if (warn) {
```

**Review by nabeelsaabna — APPROVED** [review]

**Review by rarkins — APPROVED** [review]

**Pull request merged**